src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/UnusedTrackerWithR8.groovy

```diff
@@ -0,0 +1,265 @@
+
+        this.projectFiles = []
+        this.dependencies = []
+
+        for (File dir : classDirs) {
+            Path path = Paths.get(dir.getAbsolutePath())
+            collectClassFiles(path, projectFiles)
+        }
+
+        for (File jar : classJars.getFiles()) {
+            projectFiles.add(jar)
+        }
+    }
+
+    @Override
+    @CompileStatic
+    boolean performsFullShrinking() {
+        return true
+    }
+
+    @Override
+    @CompileStatic
+    boolean hasProcessedClass(String classname) {
+        Path path = getPathToProcessedClass(classname)
+        return Files.exists(path)
+    }
+
+    @Override
+    @CompileStatic
+    Path getPathToProcessedClass(String classname) {
+        final String className = FilenameUtils.removeExtension(classname).replace('/' as char, '.' as char)
+        Path processedFile = Paths.get(tmpDir.toString(), className.replaceAll("\\.", "/") + ".class")
+        return processedFile
+    }
+
+    @Override
+    @CompileStatic
+    Set<String> findUnused() {
+        // We effectively run R8 twice:
+        //  * first time disabling any processing, to retrieve all project classes
+        //  * second time with a full shrink run to get all unused classes
+        R8Command.Builder builder = R8Command.builder()
+
+        populateBuilderWithProjectFiles(builder)
+
+        // add all dependency jars
+        for (File dep : dependencies) {
+            Path path = Paths.get(dep.getAbsolutePath())
+            builder.addProgramFiles(path)
+        }
+
+        final Set<String> removedClasses = new HashSet<>()
+
+        // Add any class from the usage list to the list of removed classes.
+        // This is a bit of a hack but the best things I could think of so far.
+        builder.setProguardUsageConsumer(new StringConsumer() {
+            private final String LINE_SEPARATOR = StringUtils.LINE_SEPARATOR
+            private String  lastString = LINE_SEPARATOR
+            private String  classString
+            private boolean expectingSeparator
+
+            @Override
+            void accept(String s, DiagnosticsHandler handler) {
+                if (classString == null && lastString == LINE_SEPARATOR) {
+                    classString        = s
+                    expectingSeparator = true
+                } else if (expectingSeparator && s == LINE_SEPARATOR) {
+                    removedClasses.add(classString)
+                    classString        = null
+                    expectingSeparator = false
+                } else {
+                    classString        = null
+                    expectingSeparator = false
+                }
+
+                lastString = s
+            }
+        })
+
+        List<String> proguardConfig = getKeepRules()
+        proguardConfig.add("-dontoptimize")
+        proguardConfig.add("-dontobfuscate")
+        proguardConfig.add("-ignorewarnings")
+
+        builder.addProguardConfiguration(proguardConfig, Origin.unknown())
```

Comment:
I will suggest that you use debug mode as well `builder.setMode(CompilationMode.DEBUG)` and not rely on `-dontoptimize` to turn off structural optimizations.

Reply:
Thinking about this again this should not use debug mode. Shouldn't `-dontoptimize` and `-dontobfuscate` be removed here? There are `-keep` rules for all classes in the shadowed jar, and the dependent classes could be renamed with a `-repackageclasses` or `-flattenpackagehierachy`. That will of course require collecting (multiple) mapping files.

Reply:
the repackaging is performed by the shadow jar plugin itself using its remapping functionality which is more powerful than the -repackageclasses. I basically want to disable any feature of R8 apart from shrinking. If you have better ideas to achieve that they are of course welcome.